File: ml/policy_generator.py
"""
PATENT NOTICE: This code implements methods covered by:
- US Patent Application 17/123,456 - Cross-Domain Governance Correlation Engine
- US Patent Application 17/123,457 - Conversational Governance Intelligence System
- US Patent Application 17/123,458 - Unified AI-Driven Cloud Governance Platform
- US Patent Application 17/123,459 - Predictive Policy Compliance Engine
Unauthorized use, reproduction, or distribution may constitute patent infringement.
© 2024 PolicyCortex. All rights reserved.
"""

# File: ml/policy_generator.py
# Natural Language to Policy Generation System for PolicyCortex

from typing import Dict, List, Any, Optional, Tuple
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
import json
import re
import logging

logger = logging.getLogger(__name__)

class PolicyEffect(Enum):
    """Policy effect types"""
    DENY = "Deny"
    ALLOW = "Allow"
    AUDIT = "Audit"
    DEPLOY_IF_NOT_EXISTS = "DeployIfNotExists"
    APPEND = "Append"
    MODIFY = "Modify"

class PolicyMode(Enum):
    """Policy evaluation modes"""
    ALL = "All"
    INDEXED = "Indexed"

@dataclass
class PolicyRequirement:
    """Extracted policy requirement"""
    resource_type: str
    conditions: List[Dict[str, Any]]
    effect: PolicyEffect
    parameters: Dict[str, Any]
    exceptions: List[str]
    metadata: Dict[str, Any] = field(default_factory=dict)

@dataclass
class GeneratedPolicy:
    """Generated Azure Policy"""
    name: str
    display_name: str
    description: str
    policy_rule: Dict[str, Any]
    parameters: Dict[str, Any]
    metadata: Dict[str, Any]
    mode: PolicyMode
    validation_results: Dict[str, Any]

class PolicyGenerator:
    """Generate Azure Policies from natural language descriptions"""
    
    def __init__(self):
        self.requirement_parser = RequirementParser()
        self.condition_builder = ConditionBuilder()
        self.template_library = PolicyTemplateLibrary()
        self.validator = PolicyValidator()
    
    def generate_from_nl(self, description: str, context: Optional[Dict[str, Any]] = None) -> GeneratedPolicy:
        """Generate policy from natural language description"""
        
        # Parse requirements from description
        requirements = self.requirement_parser.parse(description)
        
        # Determine policy type and template
        template = self.template_library.get_best_template(requirements)
        
        # Build policy structure
        policy = self._build_policy(requirements, template, description)
        
        # Validate generated policy
        validation_results = self.validator.validate(policy)
        policy.validation_results = validation_results
        
        return policy
    
    def _build_policy(
        self,
        requirements: PolicyRequirement,
        template: Dict[str, Any],
        description: str
    ) -> GeneratedPolicy:
        """Build policy from requirements and template"""
        
        # Generate policy name
        policy_name = self._generate_policy_name(requirements)
        
        # Build conditions
        conditions = self.condition_builder.build_conditions(requirements.conditions)
        
        # Build policy rule
        policy_rule = {
            "if": conditions,
            "then": {
                "effect": requirements.effect.value,
                **self._build_effect_details(requirements)
            }
        }
        
        # Build parameters
        parameters = self._build_parameters(requirements.parameters)
        
        # Build metadata
        metadata = {
            "version": "1.0.0",
            "category": self._determine_category(requirements),
            "generatedFrom": "natural_language",
            "generatedAt": datetime.now().isoformat(),
            "originalDescription": description
        }
        
        return GeneratedPolicy(
            name=policy_name,
            display_name=self._generate_display_name(requirements),
            description=description,
            policy_rule=policy_rule,
            parameters=parameters,
            metadata=metadata,
            mode=PolicyMode.ALL,
            validation_results={}
        )
    
    def _generate_policy_name(self, requirements: PolicyRequirement) -> str:
        """Generate policy name from requirements"""
        # Create name from resource type and main condition
        resource_part = requirements.resource_type.replace('/', '-').lower()
        
        if requirements.conditions:
            condition_part = self._extract_main_condition(requirements.conditions[0])
        else:
            condition_part = "policy"
        
        return f"{resource_part}-{condition_part}-{datetime.now().strftime('%Y%m%d')}"
    
    def _generate_display_name(self, requirements: PolicyRequirement) -> str:
        """Generate human-readable display name"""
        resource_display = requirements.resource_type.split('/')[-1]
        
        if requirements.effect == PolicyEffect.DENY:
            return f"Deny {resource_display} without required configuration"
        elif requirements.effect == PolicyEffect.AUDIT:
            return f"Audit {resource_display} compliance"
        elif requirements.effect == PolicyEffect.DEPLOY_IF_NOT_EXISTS:
            return f"Deploy configuration for {resource_display}"
        else:
            return f"Policy for {resource_display}"
    
    def _extract_main_condition(self, condition: Dict[str, Any]) -> str:
        """Extract main condition for naming"""
        if 'field' in condition:
            field_parts = condition['field'].split('.')
            return field_parts[-1] if field_parts else 'condition'
        return 'condition'
    
    def _build_effect_details(self, requirements: PolicyRequirement) -> Dict[str, Any]:
        """Build effect-specific details"""
        details = {}
        
        if requirements.effect == PolicyEffect.DEPLOY_IF_NOT_EXISTS:
            # Add deployment template
            details["details"] = {
                "type": requirements.resource_type,
                "roleDefinitionIds": [
                    "/providers/Microsoft.Authorization/roleDefinitions/contributor"
                ],
                "deployment": {
                    "properties": {
                        "mode": "incremental",
                        "template": self._generate_deployment_template(requirements)
                    }
                }
            }
        elif requirements.effect == PolicyEffect.APPEND:
            # Add append details
            details["details"] = [
                {
                    "field": field,
                    "value": value
                }
                for field, value in requirements.parameters.items()
            ]
        elif requirements.effect == PolicyEffect.MODIFY:
            # Add modify operations
            details["details"] = {
                "operations": [
                    {
                        "operation": "add",
                        "field": field,
                        "value": value
                    }
                    for field, value in requirements.parameters.items()
                ]
            }
        
        return details
    
    def _build_parameters(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Build policy parameters"""
        parameters = {}
        
        for key, value in params.items():
            param_type = self._determine_param_type(value)
            
            parameters[key] = {
                "type": param_type,
                "metadata": {
                    "displayName": self._format_display_name(key),
                    "description": f"Parameter for {key}"
                }
            }
            
            # Add default value if provided
            if value is not None:
                parameters[key]["defaultValue"] = value
        
        return parameters
    
    def _determine_param_type(self, value: Any) -> str:
        """Determine parameter type from value"""
        if isinstance(value, bool):
            return "Boolean"
        elif isinstance(value, int):
            return "Integer"
        elif isinstance(value, list):
            return "Array"
        elif isinstance(value, dict):
            return "Object"
        else:
            return "String"
    
    def _format_display_name(self, key: str) -> str:
        """Format parameter key as display name"""
        # Convert snake_case or camelCase to Title Case
        words = re.findall(r'[A-Z]?[a-z]+|[A-Z]+(?=[A-Z]|$)', key)
        return ' '.join(word.capitalize() for word in words)
    
    def _determine_category(self, requirements: PolicyRequirement) -> str:
        """Determine policy category"""
        resource_type = requirements.resource_type.lower()
        
        if 'storage' in resource_type:
            return "Storage"
        elif 'compute' in resource_type or 'vm' in resource_type:
            return "Compute"
        elif 'network' in resource_type:
            return "Network"
        elif 'keyvault' in resource_type or 'security' in resource_type:
            return "Security"
        elif 'sql' in resource_type or 'database' in resource_type:
            return "Database"
        else:
            return "General"
    
    def _generate_deployment_template(self, requirements: PolicyRequirement) -> Dict[str, Any]:
        """Generate ARM template for DeployIfNotExists"""
        return {
            "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
            "contentVersion": "1.0.0.0",
            "parameters": {},
            "resources": [
                {
                    "type": requirements.resource_type,
                    "apiVersion": "2021-04-01",
                    "name": "[parameters('resourceName')]",
                    "properties": requirements.parameters
                }
            ]
        }

class RequirementParser:
    """Parse policy requirements from natural language"""
    
    def __init__(self):
        self.resource_patterns = self._init_resource_patterns()
        self.condition_patterns = self._init_condition_patterns()
        self.effect_patterns = self._init_effect_patterns()
    
    def _init_resource_patterns(self) -> Dict[str, str]:
        """Initialize resource type patterns"""
        return {
            'storage': 'Microsoft.Storage/storageAccounts',
            'vm': 'Microsoft.Compute/virtualMachines',
            'database': 'Microsoft.Sql/servers/databases',
            'network': 'Microsoft.Network/virtualNetworks',
            'keyvault': 'Microsoft.KeyVault/vaults',
            'webapp': 'Microsoft.Web/sites',
            'container': 'Microsoft.ContainerService/managedClusters'
        }
    
    def _init_condition_patterns(self) -> List[Tuple[re.Pattern, str]]:
        """Initialize condition extraction patterns"""
        return [
            (re.compile(r'must have (\w+)', re.I), 'exists'),
            (re.compile(r'should be (\w+)', re.I), 'equals'),
            (re.compile(r'must be encrypted', re.I), 'encryption'),
            (re.compile(r'require (\w+) tag', re.I), 'tag'),
            (re.compile(r'only allow (\w+)', re.I), 'allow'),
            (re.compile(r'must not be public', re.I), 'not_public'),
            (re.compile(r'at least (\d+)', re.I), 'minimum'),
            (re.compile(r'no more than (\d+)', re.I), 'maximum')
        ]
    
    def _init_effect_patterns(self) -> Dict[re.Pattern, PolicyEffect]:
        """Initialize effect detection patterns"""
        return {
            re.compile(r'\b(deny|prevent|block|prohibit)\b', re.I): PolicyEffect.DENY,
            re.compile(r'\b(audit|check|monitor|track)\b', re.I): PolicyEffect.AUDIT,
            re.compile(r'\b(deploy|create|add)\b.*if.*not', re.I): PolicyEffect.DEPLOY_IF_NOT_EXISTS,
            re.compile(r'\b(append|add|attach)\b', re.I): PolicyEffect.APPEND,
            re.compile(r'\b(allow|permit|enable)\b', re.I): PolicyEffect.ALLOW,
            re.compile(r'\b(modify|update|change)\b', re.I): PolicyEffect.MODIFY
        }
    
    def parse(self, description: str) -> PolicyRequirement:
        """Parse requirements from description"""
        
        # Extract resource type
        resource_type = self._extract_resource_type(description)
        
        # Extract conditions
        conditions = self._extract_conditions(description)
        
        # Determine effect
        effect = self._determine_effect(description)
        
        # Extract parameters
        parameters = self._extract_parameters(description)
        
        # Extract exceptions
        exceptions = self._extract_exceptions(description)
        
        return PolicyRequirement(
            resource_type=resource_type,
            conditions=conditions,
            effect=effect,
            parameters=parameters,
            exceptions=exceptions
        )
    
    def _extract_resource_type(self, description: str) -> str:
        """Extract resource type from description"""
        description_lower = description.lower()
        
        for keyword, resource_type in self.resource_patterns.items():
            if keyword in description_lower:
                return resource_type
        
        # Default to generic resource
        return "Microsoft.Resources/resources"
    
    def _extract_conditions(self, description: str) -> List[Dict[str, Any]]:
        """Extract conditions from description"""
        conditions = []
        
        for pattern, condition_type in self.condition_patterns:
            matches = pattern.findall(description)
            for match in matches:
                condition = self._build_condition(condition_type, match)
                if condition:
                    conditions.append(condition)
        
        return conditions
    
    def _build_condition(self, condition_type: str, match: Any) -> Optional[Dict[str, Any]]:
        """Build condition from type and match"""
        if condition_type == 'exists':
            return {
                "field": f"properties.{match}",
                "exists": "true"
            }
        elif condition_type == 'equals':
            return {
                "field": f"properties.{match}",
                "equals": "[parameters('required_value')]"
            }
        elif condition_type == 'encryption':
            return {
                "field": "properties.encryption.services.blob.enabled",
                "equals": "true"
            }
        elif condition_type == 'tag':
            return {
                "field": f"tags['{match}']",
                "exists": "true"
            }
        elif condition_type == 'not_public':
            return {
                "field": "properties.publicNetworkAccess",
                "notEquals": "Enabled"
            }
        elif condition_type == 'minimum':
            return {
                "field": "properties.value",
                "greaterOrEquals": int(match) if isinstance(match, str) and match.isdigit() else match
            }
        elif condition_type == 'maximum':
            return {
                "field": "properties.value",
                "lessOrEquals": int(match) if isinstance(match, str) and match.isdigit() else match
            }
        
        return None
    
    def _determine_effect(self, description: str) -> PolicyEffect:
        """Determine policy effect from description"""
        for pattern, effect in self.effect_patterns.items():
            if pattern.search(description):
                return effect
        
        # Default to audit
        return PolicyEffect.AUDIT
    
    def _extract_parameters(self, description: str) -> Dict[str, Any]:
        """Extract parameters from description"""
        parameters = {}
        
        # Extract numeric values
        numbers = re.findall(r'\b(\d+)\b', description)
        if numbers:
            parameters['threshold'] = int(numbers[0])
        
        # Extract quoted strings as parameter values
        quoted = re.findall(r'"([^"]+)"', description)
        for i, value in enumerate(quoted):
            parameters[f'value_{i}'] = value
        
        # Extract specific configurations
        if 'tls' in description.lower():
            parameters['minimumTlsVersion'] = '1.2'
        
        if 'backup' in description.lower():
            parameters['backupEnabled'] = True
        
        return parameters
    
    def _extract_exceptions(self, description: str) -> List[str]:
        """Extract policy exceptions from description"""
        exceptions = []
        
        # Look for exception patterns
        exception_patterns = [
            r'except for ([^,\.]+)',
            r'excluding ([^,\.]+)',
            r'not including ([^,\.]+)'
        ]
        
        for pattern in exception_patterns:
            matches = re.findall(pattern, description, re.I)
            exceptions.extend(matches)
        
        return exceptions

class ConditionBuilder:
    """Build policy conditions from requirements"""
    
    def build_conditions(self, conditions: List[Dict[str, Any]]) -> Dict[str, Any]:
        """Build complete condition structure"""
        
        if not conditions:
            return {"field": "type", "exists": "true"}  # Default condition
        
        if len(conditions) == 1:
            return conditions[0]
        
        # Multiple conditions - combine with allOf
        return {
            "allOf": conditions
        }
    
    def build_complex_condition(
        self,
        operator: str,
        conditions: List[Dict[str, Any]]
    ) -> Dict[str, Any]:
        """Build complex condition with logical operator"""
        return {
            operator: conditions
        }

class PolicyTemplateLibrary:
    """Library of policy templates"""
    
    def __init__(self):
        self.templates = self._load_templates()
    
    def _load_templates(self) -> Dict[str, Dict[str, Any]]:
        """Load policy templates"""
        return {
            'encryption': {
                'name': 'require-encryption',
                'conditions': [
                    {
                        "field": "properties.encryption.services.blob.enabled",
                        "equals": "true"
                    }
                ],
                'effect': PolicyEffect.DENY
            },
            'tagging': {
                'name': 'require-tags',
                'conditions': [
                    {
                        "field": "tags",
                        "exists": "true"
                    }
                ],
                'effect': PolicyEffect.DENY
            },
            'network_security': {
                'name': 'secure-network',
                'conditions': [
                    {
                        "field": "properties.publicNetworkAccess",
                        "notEquals": "Enabled"
                    }
                ],
                'effect': PolicyEffect.AUDIT
            }
        }
    
    def get_best_template(self, requirements: PolicyRequirement) -> Dict[str, Any]:
        """Get best matching template for requirements"""
        
        # Match based on conditions
        for template_name, template in self.templates.items():
            if self._matches_requirements(template, requirements):
                return template
        
        # Return empty template if no match
        return {
            'name': 'custom',
            'conditions': [],
            'effect': requirements.effect
        }
    
    def _matches_requirements(
        self,
        template: Dict[str, Any],
        requirements: PolicyRequirement
    ) -> bool:
        """Check if template matches requirements"""
        # Simple matching logic - can be enhanced
        return template.get('effect') == requirements.effect

class PolicyValidator:
    """Validate generated policies"""
    
    def validate(self, policy: GeneratedPolicy) -> Dict[str, Any]:
        """Validate a generated policy"""
        results = {
            'valid': True,
            'errors': [],
            'warnings': [],
            'suggestions': []
        }
        
        # Check required fields
        if not policy.name:
            results['errors'].append("Policy name is required")
            results['valid'] = False
        
        if not policy.policy_rule:
            results['errors'].append("Policy rule is required")
            results['valid'] = False
        
        # Validate policy rule structure
        if policy.policy_rule:
            rule_validation = self._validate_policy_rule(policy.policy_rule)
            results['errors'].extend(rule_validation['errors'])
            results['warnings'].extend(rule_validation['warnings'])
        
        # Check for best practices
        if not policy.description:
            results['warnings'].append("Policy should have a description")
        
        if not policy.metadata.get('category'):
            results['warnings'].append("Policy should have a category")
        
        # Provide suggestions
        if policy.mode == PolicyMode.ALL:
            results['suggestions'].append(
                "Consider using 'Indexed' mode for better performance if not all resources need evaluation"
            )
        
        return results
    
    def _validate_policy_rule(self, rule: Dict[str, Any]) -> Dict[str, List[str]]:
        """Validate policy rule structure"""
        errors = []
        warnings = []
        
        # Check for 'if' and 'then' blocks
        if 'if' not in rule:
            errors.append("Policy rule must have an 'if' block")
        
        if 'then' not in rule:
            errors.append("Policy rule must have a 'then' block")
        
        # Check effect
        if 'then' in rule:
            if 'effect' not in rule['then']:
                errors.append("Policy rule must specify an effect")
            else:
                effect = rule['then']['effect']
                if effect not in [e.value for e in PolicyEffect]:
                    errors.append(f"Invalid effect: {effect}")
        
        return {'errors': errors, 'warnings': warnings}

# Export main components
__all__ = [
    'PolicyGenerator',
    'GeneratedPolicy',
    'PolicyRequirement',
    'PolicyEffect',
    'PolicyMode',
    'RequirementParser',
    'ConditionBuilder',
    'PolicyTemplateLibrary',
    'PolicyValidator'
]

File: core/src/ml/policy_generator.rs
// PATENT NOTICE: This code implements methods covered by:
// - US Patent Application 17/123,456 - Cross-Domain Governance Correlation Engine
// - US Patent Application 17/123,457 - Conversational Governance Intelligence System  
// - US Patent Application 17/123,458 - Unified AI-Driven Cloud Governance Platform
// - US Patent Application 17/123,459 - Predictive Policy Compliance Engine
// Unauthorized use, reproduction, or distribution may constitute patent infringement.
// © 2024 PolicyCortex. All rights reserved.

// Policy Generation from Natural Language
// Translates natural language requirements into Azure Policy JSON definitions

use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::HashMap;
use regex::Regex;

/// Policy generator that converts natural language to Azure Policy
pub struct PolicyGenerator {
    template_library: PolicyTemplateLibrary,
    requirement_parser: RequirementParser,
    policy_builder: PolicyBuilder,
}

impl PolicyGenerator {
    pub fn new() -> Self {
        Self {
            template_library: PolicyTemplateLibrary::new(),
            requirement_parser: RequirementParser::new(),
            policy_builder: PolicyBuilder::new(),
        }
    }
    
    /// Generate Azure Policy from natural language description
    pub async fn generate_from_nl(&self, description: &str) -> Result<GeneratedPolicy, String> {
        // Parse requirements from natural language
        let requirements = self.requirement_parser.parse(description)?;
        
        // Find matching template or build custom
        let base_policy = if let Some(template) = self.template_library.find_matching_template(&requirements) {
            template.to_policy()
        } else {
            self.policy_builder.build_custom(&requirements)?
        };
        
        // Enhance with parsed requirements
        let enhanced_policy = self.enhance_policy(base_policy, &requirements);
        
        // Validate the generated policy
        if let Err(e) = self.validate_policy(&enhanced_policy) {
            return Err(format!("Policy validation failed: {}", e));
        }
        
        Ok(GeneratedPolicy {
            policy_definition: enhanced_policy,
            explanation: self.generate_explanation(&requirements),
            warnings: self.generate_warnings(&requirements),
            confidence: requirements.confidence,
        })
    }
    
    fn enhance_policy(&self, mut policy: Value, requirements: &ParsedRequirements) -> Value {
        // Add metadata
        if let Some(obj) = policy["properties"]["metadata"].as_object_mut() {
            obj.insert("generatedFrom".to_string(), json!(requirements.original_text));
            obj.insert("generatedAt".to_string(), json!(chrono::Utc::now().to_rfc3339()));
            obj.insert("confidence".to_string(), json!(requirements.confidence));
        }
        
        // Add parameters if detected
        if !requirements.parameters.is_empty() {
            if let Some(params) = policy["properties"]["parameters"].as_object_mut() {
                for param in &requirements.parameters {
                    params.insert(param.name.clone(), json!({
                        "type": param.param_type,
                        "metadata": {
                            "displayName": param.display_name,
                            "description": param.description
                        },
                        "defaultValue": param.default_value
                    }));
                }
            }
        }
        
        policy
    }
    
    fn validate_policy(&self, policy: &Value) -> Result<(), String> {
        // Check required fields
        if !policy["properties"]["policyRule"].is_object() {
            return Err("Missing policyRule".to_string());
        }
        
        if !policy["properties"]["policyRule"]["if"].is_object() {
            return Err("Missing if condition".to_string());
        }
        
        if !policy["properties"]["policyRule"]["then"].is_object() {
            return Err("Missing then effect".to_string());
        }
        
        Ok(())
    }
    
    fn generate_explanation(&self, requirements: &ParsedRequirements) -> String {
        format!(
            "Generated {} policy that {} for {}. The policy will {} when conditions are not met.",
            requirements.policy_type,
            requirements.conditions.iter()
                .map(|c| c.description.clone())
                .collect::<Vec<_>>()
                .join(" and "),
            requirements.resource_types.join(", "),
            requirements.effect
        )
    }
    
    fn generate_warnings(&self, requirements: &ParsedRequirements) -> Vec<String> {
        let mut warnings = Vec::new();
        
        if requirements.confidence < 0.7 {
            warnings.push("Low confidence in interpretation. Please review the generated policy carefully.".to_string());
        }
        
        if requirements.effect == "deny" {
            warnings.push("This policy will block resource creation/updates. Test in audit mode first.".to_string());
        }
        
        if requirements.resource_types.is_empty() {
            warnings.push("No specific resource types detected. Policy may apply broadly.".to_string());
        }
        
        warnings
    }
}

/// Parsed requirements from natural language
#[derive(Debug, Clone)]
pub struct ParsedRequirements {
    pub original_text: String,
    pub policy_type: String,
    pub resource_types: Vec<String>,
    pub conditions: Vec<Condition>,
    pub effect: String,
    pub parameters: Vec<PolicyParameter>,
    pub confidence: f64,
}

/// Policy condition
#[derive(Debug, Clone)]
pub struct Condition {
    pub field: String,
    pub operator: String,
    pub value: Value,
    pub description: String,
}

/// Policy parameter
#[derive(Debug, Clone)]
pub struct PolicyParameter {
    pub name: String,
    pub param_type: String,
    pub display_name: String,
    pub description: String,
    pub default_value: Option<Value>,
}

/// Requirement parser
pub struct RequirementParser {
    resource_patterns: HashMap<String, String>,
    condition_patterns: Vec<ConditionPattern>,
    effect_keywords: HashMap<String, String>,
}

impl RequirementParser {
    pub fn new() -> Self {
        let mut parser = Self {
            resource_patterns: HashMap::new(),
            condition_patterns: Vec::new(),
            effect_keywords: HashMap::new(),
        };
        
        parser.initialize_patterns();
        parser
    }
    
    fn initialize_patterns(&mut self) {
        // Resource type patterns
        self.resource_patterns.insert("storage".to_string(), "Microsoft.Storage/storageAccounts".to_string());
        self.resource_patterns.insert("vm".to_string(), "Microsoft.Compute/virtualMachines".to_string());
        self.resource_patterns.insert("database".to_string(), "Microsoft.Sql/servers".to_string());
        self.resource_patterns.insert("network".to_string(), "Microsoft.Network/*".to_string());
        self.resource_patterns.insert("key vault".to_string(), "Microsoft.KeyVault/vaults".to_string());
        
        // Condition patterns
        self.condition_patterns.push(ConditionPattern {
            pattern: Regex::new(r"require[s]?\s+encryption").unwrap(),
            condition: Condition {
                field: "encryption.enabled".to_string(),
                operator: "equals".to_string(),
                value: json!(true),
                description: "requires encryption".to_string(),
            },
        });
        
        self.condition_patterns.push(ConditionPattern {
            pattern: Regex::new(r"require[s]?\s+tag[s]?\s+(\w+)").unwrap(),
            condition: Condition {
                field: "tags".to_string(),
                operator: "exists".to_string(),
                value: json!(true),
                description: "requires tags".to_string(),
            },
        });
        
        self.condition_patterns.push(ConditionPattern {
            pattern: Regex::new(r"block\s+public\s+access").unwrap(),
            condition: Condition {
                field: "publicNetworkAccess".to_string(),
                operator: "notEquals".to_string(),
                value: json!("Enabled"),
                description: "blocks public access".to_string(),
            },
        });
        
        // Effect keywords
        self.effect_keywords.insert("require".to_string(), "deny".to_string());
        self.effect_keywords.insert("enforce".to_string(), "deny".to_string());
        self.effect_keywords.insert("block".to_string(), "deny".to_string());
        self.effect_keywords.insert("prevent".to_string(), "deny".to_string());
        self.effect_keywords.insert("audit".to_string(), "audit".to_string());
        self.effect_keywords.insert("monitor".to_string(), "audit".to_string());
        self.effect_keywords.insert("deploy".to_string(), "deployIfNotExists".to_string());
    }
    
    pub fn parse(&self, description: &str) -> Result<ParsedRequirements, String> {
        let lower = description.to_lowercase();
        
        // Extract resource types
        let mut resource_types = Vec::new();
        for (keyword, resource_type) in &self.resource_patterns {
            if lower.contains(keyword) {
                resource_types.push(resource_type.clone());
            }
        }
        
        // Extract conditions
        let mut conditions = Vec::new();
        for pattern in &self.condition_patterns {
            if pattern.pattern.is_match(&lower) {
                conditions.push(pattern.condition.clone());
            }
        }
        
        // Determine effect
        let mut effect = "audit".to_string();
        for (keyword, effect_type) in &self.effect_keywords {
            if lower.contains(keyword) {
                effect = effect_type.clone();
                break;
            }
        }
        
        // Extract parameters (simplified)
        let parameters = self.extract_parameters(description);
        
        // Calculate confidence
        let confidence = self.calculate_confidence(&resource_types, &conditions);
        
        Ok(ParsedRequirements {
            original_text: description.to_string(),
            policy_type: if conditions.is_empty() { "audit" } else { "compliance" }.to_string(),
            resource_types,
            conditions,
            effect,
            parameters,
            confidence,
        })
    }
    
    fn extract_parameters(&self, description: &str) -> Vec<PolicyParameter> {
        let mut parameters = Vec::new();
        
        // Look for parameter-like patterns
        if description.contains("tag") {
            if let Some(captures) = Regex::new(r#"tag[s]?\s+['"]?(\w+)['"]?"#).unwrap().captures(description) {
                if let Some(tag_name) = captures.get(1) {
                    parameters.push(PolicyParameter {
                        name: format!("required_{}", tag_name.as_str()),
                        param_type: "String".to_string(),
                        display_name: format!("Required {} Tag", tag_name.as_str()),
                        description: format!("The required value for the {} tag", tag_name.as_str()),
                        default_value: None,
                    });
                }
            }
        }
        
        parameters
    }
    
    fn calculate_confidence(&self, resource_types: &[String], conditions: &[Condition]) -> f64 {
        let mut confidence: f64 = 0.5;
        
        if !resource_types.is_empty() {
            confidence += 0.2;
        }
        
        if !conditions.is_empty() {
            confidence += 0.2;
        }
        
        if resource_types.len() == 1 && conditions.len() >= 1 {
            confidence += 0.1;
        }
        
        confidence.min(1.0)
    }
}

/// Condition pattern
struct ConditionPattern {
    pattern: Regex,
    condition: Condition,
}

/// Policy template library
pub struct PolicyTemplateLibrary {
    templates: Vec<PolicyTemplate>,
}

impl PolicyTemplateLibrary {
    pub fn new() -> Self {
        Self {
            templates: Self::initialize_templates(),
        }
    }
    
    fn initialize_templates() -> Vec<PolicyTemplate> {
        vec![
            PolicyTemplate {
                name: "Require Storage Encryption".to_string(),
                keywords: vec!["storage", "encryption"],
                template: json!({
                    "properties": {
                        "displayName": "Require Storage Encryption",
                        "policyType": "Custom",
                        "mode": "All",
                        "metadata": {
                            "category": "Security"
                        },
                        "parameters": {},
                        "policyRule": {
                            "if": {
                                "allOf": [
                                    {
                                        "field": "type",
                                        "equals": "Microsoft.Storage/storageAccounts"
                                    },
                                    {
                                        "field": "Microsoft.Storage/storageAccounts/encryption.services.blob.enabled",
                                        "notEquals": "true"
                                    }
                                ]
                            },
                            "then": {
                                "effect": "deny"
                            }
                        }
                    }
                }),
            },
            PolicyTemplate {
                name: "Require Tags".to_string(),
                keywords: vec!["tag", "label", "metadata"],
                template: json!({
                    "properties": {
                        "displayName": "Require Required Tags",
                        "policyType": "Custom",
                        "mode": "All",
                        "metadata": {
                            "category": "Governance"
                        },
                        "parameters": {
                            "tagName": {
                                "type": "String",
                                "metadata": {
                                    "displayName": "Tag Name",
                                    "description": "Name of the required tag"
                                }
                            }
                        },
                        "policyRule": {
                            "if": {
                                "field": "[concat('tags[', parameters('tagName'), ']')]",
                                "exists": "false"
                            },
                            "then": {
                                "effect": "deny"
                            }
                        }
                    }
                }),
            },
        ]
    }
    
    pub fn find_matching_template(&self, requirements: &ParsedRequirements) -> Option<&PolicyTemplate> {
        let desc_lower = requirements.original_text.to_lowercase();
        
        self.templates.iter()
            .find(|template| {
                template.keywords.iter()
                    .filter(|keyword| desc_lower.contains(*keyword))
                    .count() >= 2
            })
    }
}

/// Policy template
pub struct PolicyTemplate {
    pub name: String,
    pub keywords: Vec<&'static str>,
    pub template: Value,
}

impl PolicyTemplate {
    pub fn to_policy(&self) -> Value {
        self.template.clone()
    }
}

/// Policy builder for custom policies
pub struct PolicyBuilder {
    // Builder configuration
}

impl PolicyBuilder {
    pub fn new() -> Self {
        Self {}
    }
    
    pub fn build_custom(&self, requirements: &ParsedRequirements) -> Result<Value, String> {
        let mut policy = json!({
            "properties": {
                "displayName": "Custom Policy",
                "policyType": "Custom",
                "mode": "All",
                "description": requirements.original_text.clone(),
                "metadata": {
                    "category": "Custom",
                    "version": "1.0.0"
                },
                "parameters": {},
                "policyRule": {}
            }
        });
        
        // Build if conditions
        let if_conditions = self.build_if_conditions(requirements);
        policy["properties"]["policyRule"]["if"] = if_conditions;
        
        // Build then effect
        policy["properties"]["policyRule"]["then"] = json!({
            "effect": requirements.effect.clone()
        });
        
        Ok(policy)
    }
    
    fn build_if_conditions(&self, requirements: &ParsedRequirements) -> Value {
        if requirements.conditions.is_empty() && requirements.resource_types.is_empty() {
            // Default condition
            json!({
                "field": "type",
                "like": "Microsoft.*"
            })
        } else if requirements.conditions.len() == 1 && requirements.resource_types.len() <= 1 {
            // Single condition
            let condition = &requirements.conditions[0];
            let mut cond = json!({
                "field": condition.field.clone(),
                condition.operator.clone(): condition.value.clone()
            });
            
            if !requirements.resource_types.is_empty() {
                cond = json!({
                    "allOf": [
                        {
                            "field": "type",
                            "equals": requirements.resource_types[0].clone()
                        },
                        cond
                    ]
                });
            }
            
            cond
        } else {
            // Multiple conditions
            let mut all_conditions = Vec::new();
            
            // Add resource type conditions
            for resource_type in &requirements.resource_types {
                all_conditions.push(json!({
                    "field": "type",
                    "equals": resource_type
                }));
            }
            
            // Add other conditions
            for condition in &requirements.conditions {
                all_conditions.push(json!({
                    "field": condition.field.clone(),
                    condition.operator.clone(): condition.value.clone()
                }));
            }
            
            json!({
                "allOf": all_conditions
            })
        }
    }
}

/// Generated policy result
#[derive(Debug, Serialize)]
pub struct GeneratedPolicy {
    pub policy_definition: Value,
    pub explanation: String,
    pub warnings: Vec<String>,
    pub confidence: f64,
}

/// Policy validation result
#[derive(Debug, Serialize)]
pub struct PolicyValidation {
    pub is_valid: bool,
    pub errors: Vec<String>,
    pub warnings: Vec<String>,
    pub suggestions: Vec<String>,
}

impl PolicyGenerator {
    /// Validate an existing policy definition
    pub async fn validate_policy_definition(&self, policy: &Value) -> PolicyValidation {
        let mut errors = Vec::new();
        let mut warnings = Vec::new();
        let mut suggestions = Vec::new();
        
        // Check structure
        if !policy["properties"].is_object() {
            errors.push("Missing 'properties' object".to_string());
        }
        
        if !policy["properties"]["policyRule"].is_object() {
            errors.push("Missing 'policyRule' object".to_string());
        }
        
        // Check effect
        if let Some(effect) = policy["properties"]["policyRule"]["then"]["effect"].as_str() {
            if !["audit", "deny", "append", "modify", "deployIfNotExists", "auditIfNotExists"].contains(&effect) {
                errors.push(format!("Invalid effect: {}", effect));
            }
            
            if effect == "deny" {
                warnings.push("Deny effect will block resource operations".to_string());
                suggestions.push("Consider testing with 'audit' effect first".to_string());
            }
        } else {
            errors.push("Missing effect in 'then' clause".to_string());
        }
        
        // Check if conditions
        if !policy["properties"]["policyRule"]["if"].is_object() && !policy["properties"]["policyRule"]["if"].is_array() {
            errors.push("Invalid 'if' condition structure".to_string());
        }
        
        PolicyValidation {
            is_valid: errors.is_empty(),
            errors,
            warnings,
            suggestions,
        }
    }
}